### PR TITLE
Add support for comment node

### LINKED ISSUE #10: Support comment node
Parsing `padding:/*before*/ 2px /*between*/ 1rem /*after*/;` gives:

```js
{ type: 'word', value: '2px' }
{ type: 'div', value: '/', before: ' ', after: '' }
{ type: 'word', value: '*between*' }
{ type: 'div', value: '/', before: '', after: ' ' }
{ type: 'word', value: '1rem' }
{ type: 'div', value: '/', before: ' ', after: '' }
{ type: 'word', value: '*after*' }
{ type: 'div', value: '/', before: '', after: '' }
```

I think it makes more sense to have a comment node.

Thanks!

### PR DESCRIPTION
**Pull request by MohammadYounes**
Closes #10

### CODE REVIEW

**Comment by TrySound** [comment]
Use `value.charCodeAt(pos + 1)` directly instead of `peek` function. It's faster

**Comment by MohammadYounes** [comment]
Done!

**Comment by TrySound** [comment]
@MohammadYounes I'm not sure we should include whitespace around comment. It's part of the value content. If we will remove comment from tree then content can be broken.

**Comment by MohammadYounes** [comment]
@TrySound Ok, I'll update this later tonight.

**Comment by ben-eb** [comment]
It would be better if `/*` and `*/` didn't appear in `before` or `after` - that should be up to the stringifier. And, we should throw on unclosed comments, as PostCSS does.

https://github.com/postcss/postcss/blob/master/lib/tokenize.es6#L191-L194

**Comment by TrySound** [comment]
@MohammadYounes Let's do the same as with functions.

**Comment by MohammadYounes** [comment]
@TrySound Updated!

File: lib/parse.js [hunk 1]
```diff
@@ -38,7 +41,7 @@ module.exports = function (input) {
                 after = token;
             } else if (prev && prev.type === 'div') {
                 prev.after = token;
-            } else if (code === slash || code === comma || code === colon) {
+            } else if ((code === slash && value.charCodeAt(next + 1) !== star) || code === comma || code === colon) {
```
Comment (TrySound) [inline lib/parse.js]: && has more specifciity than || so you can remove extra parentheses.

File: lib/parse.js [hunk 2]
```diff
@@ -164,7 +183,7 @@ module.exports = function (input) {
             } while (next < max && !(
                 code <= 32 ||
                 code === singleQuote || code === doubleQuote ||
-                code === slash || code === comma || code === colon ||
+                (code === slash && value.charCodeAt(next + 1) !== star) || code === comma || code === colon ||
```
Comment (TrySound) [inline lib/parse.js]: The same, but also you can add test when you have slash before comment.

File: lib/parse.js (hunk 1 above)
Comment (MohammadYounes) [inline lib/parse.js]: Ok, Updated!

File: lib/parse.js (hunk 2 above)
Comment (MohammadYounes) [inline lib/parse.js]: @TrySound  why do we need to test for slash before comment ?

Comment (TrySound) [inline lib/parse.js]: All cases should be parsed correctly.

Comment (MohammadYounes) [inline lib/parse.js]: What's the case are you talking about ?

Comment (TrySound) [inline lib/parse.js]: Hm.. okay. Seems I was wrong. Everything should be processed correctly.

File: lib/parse.js [hunk 3]
```diff
@@ -93,7 +112,7 @@ module.exports = function (input) {
             token = { type: 'function', sourceIndex: pos - name.length, value: name, before: value.slice(pos + 1, next) };
             pos = next;
 
-            if (name === 'url' && code !== singleQuote && code !== doubleQuote) {
+            if (name === 'url' && code !== singleQuote && code !== doubleQuote && !(code === slash && value.charCodeAt(pos + 1) === star)) {
```
Comment (TrySound) [inline lib/parse.js]: Here better this I think
```js
(code !== slash || value.charCodeAt(pos + 1) !== star)
```

File: test/parse.js [hunk 4]
```diff
@@ -389,6 +389,41 @@ var tests = [{
...
+        { type: 'space', sourceIndex: 14, value: ' ' },
+        { type: 'comment', sourceIndex: 15, value: 'between'},
+        { type: 'space', sourceIndex: 26, value: ' ' },
+        { type: 'word', sourceIndex: 27, value: '1px' },
+        { type: 'space', sourceIndex: 30, value: ' ' },
+        { type: 'comment', sourceIndex: 31, value: 'after'},
+    ]
+}, {
+    message: 'should parse comments inside functions',
+    fixture: 'url( "/demo/bg.png" /*comment*/ )',
```
Comment (TrySound) [inline test/parse.js]: Add also test case with unquoted `url` arguments. Url function has different parser.

File: lib/parse.js (hunk 3 above)
Comment (MohammadYounes) [inline lib/parse.js]: Okay, I'll update. For me `!(comment start)` is more readable :)

Comment (TrySound) [inline lib/parse.js]: I just don't like ! before parentheses. More like De Morgan's laws in such cases.

File: test/parse.js (hunk 4 above)
Comment (MohammadYounes) [inline test/parse.js]: It's invalid CSS syntax! do you want to support it ?

Comment (TrySound) [inline test/parse.js]: No, but I need to know what is in output in this case. Test is the better place for this.

Comment (MohammadYounes) [inline test/parse.js]: Currently, the word parsing logic [here](https://github.com/TrySound/postcss-value-parser/blob/master/lib/parse.js#L98-L112) will eat up the comment.

Seems it wasn't supported before!

Comment (TrySound) [inline test/parse.js]: It wasn't. But it's often case. Url can contain data:uri. That's why we should optimize a little this process.

File: lib/parse.js [hunk 5]
```diff
@@ -72,6 +75,22 @@ module.exports = function (input) {
...
+        } else if (code === slash && value.charCodeAt(pos + 1) === star) {
+            token = { type: 'comment', sourceIndex: pos};
+
+            next = value.indexOf(commentEnd, pos);
+            if (next < 0) {
+                token.unclosed = true;
+                next = value.length;                
+            }
+
+            token.value = value.slice(pos + commentStart.length, next);
```
Comment (TrySound) [inline lib/parse.js]: We know that `commentStart.length` is two. It's intuitive constant. Better use comment instead of variable.

File: test/parse.js (hunk 4 above)
Comment (MohammadYounes) [inline test/parse.js]: Sure, but may be a different PR!

File: lib/parse.js [hunk 6]
```diff
@@ -72,6 +75,22 @@ module.exports = function (input) {
             pos = next + 1;
             code = value.charCodeAt(pos);
 
+        //Comments
+        } else if (code === slash && value.charCodeAt(pos + 1) === star) {
+            token = { type: 'comment', sourceIndex: pos};
+
+            next = value.indexOf(commentEnd, pos);
```
Comment (TrySound) [inline lib/parse.js]: Just add the string, not variable.

File: lib/parse.js [hunk 7]
```diff
@@ -6,6 +6,9 @@ var backslash = '\\'.charCodeAt(0);
 var slash = '/'.charCodeAt(0);
 var comma = ','.charCodeAt(0);
 var colon = ':'.charCodeAt(0);
+var star = '*'.charCodeAt(0);
+var commentStart = '/*';
+var commentEnd = '*/';
```
Comment (TrySound) [inline lib/parse.js]: These lines should be removed

File: test/parse.js (hunk 4 above)
Comment (TrySound) [inline test/parse.js]: Okay doesn't matter. I know everything will be okay :)

File: lib/parse.js (hunk 5 above)
Comment (MohammadYounes) [inline lib/parse.js]: Well, I think there is no need for comment, I was just minding readability :)

File: lib/parse.js (hunk 6 above)
Comment (MohammadYounes) [inline lib/parse.js]: Done

File: lib/parse.js (hunk 7 above)
Comment (MohammadYounes) [inline lib/parse.js]: Done

File: lib/parse.js [hunk 8]
```diff
@@ -72,6 +73,22 @@ module.exports = function (input) {
             pos = next + 1;
             code = value.charCodeAt(pos);
 
+        //Comments
+        } else if (code === slash && value.charCodeAt(pos + 1) === star) {
+            token = { type: 'comment', sourceIndex: pos};
+
+            next = value.indexOf('*/', pos);
+            if (next < 0) {
```
Comment (TrySound) [inline lib/parse.js]: `next === -1`

**Comment by TrySound** [comment]
Nice job. Thanks.

File: lib/parse.js (hunk 8 above)
Comment (MohammadYounes) [inline lib/parse.js]: Done
